### ISSUE TITLE
Proposal: Teach Promises

### ISSUE THREAD

**Issue by oliverjam** [foundersandcoders/master-reference#262]
## Why

Promises are really important to modern JavaScript, especially client-side. Nearly every new browser API is promise-based (Service Workers, Fetch etc). They also make dealing with asynchronous code much more intuitive (in my opinion).

We're currently at Week 13 and unless they've come across them in their own time nobody knows what promises are yet.

## What

I think a good place to introduce promises would be API Week. If we stop teaching the confusing and outdated `XmLHttPrequest` API and teach `fetch` instead this would be a great time to introduce the concept of promises.

The Fetch API [is supported](http://caniuse.com/#feat=fetch) in all modern browsers (and can be [polyfilled](https://github.com/github/fetch) back to IE10). Whilst  learning underlying technologies rather than abstractions is useful, I don't think that applies here. XMLHttpRequest is just a random old browser API that has been replaced with something better in every way — no one needs to use it anymore unless they're supporting IE8/9. It doesn't teach you anything fundamental except callbacks, which I think actually get confused because there's so much other cruft involved in making a request.

## How

The FAC10 curriculum didn't introduce the concept of callbacks until the Parallel Morning Challenge on Day 3. We spent two days learning to make requests without fully understanding how. I don't see how this would be any different using a promise-based API. Students would learn the syntax first and then be introduced to the underlying concepts later in the week.

I would consider moving the Parallel Morning Challenge to later in the course and replace it in this week with a Workshop covering callbacks and promises. So the structure of the week would be relatively unchanged other than amending Lucy's workshop on HTTP requests to mention XMLHttpRequest as a historical footnote before teaching the Fetch syntax, and using Fetch rather than XMLHttpRequest for the other workshops.

We'd then have a new workshop teaching async basics, including callbacks and promises. Introducing callbacks and promises in quick succession has the potential for confusion so this workshop would need to be well-structured, but I don't think it's unreasonable to cover the basics of asynchronicity in an hour.

**Comment by ronanyeah** [comment]
I wrote these for the Gaza students to teach XHR abstraction and promises, before I started recommending fetch. I think they could be useful for the curriculum.
https://github.com/ronanyeah/xhr-wrapper-workshop
https://github.com/ronanyeah/promise-workshop

**Comment by ronanyeah** [comment]
Fetch is also quite low level and has several 'gotchas' ( passing cookies being a notable one) but it much nicer to use that XMLHttpRequest, the grossest API of all time.

https://medium.com/@shahata/why-i-wont-be-using-fetch-api-in-my-apps-6900e6c6fe78

**Comment by ronanyeah** [comment]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

Async/await is also in Node.js core, which is a great way to structure your asynchronous code.

These are all tools being used in [modern JavaScript](https://github.com/yarnpkg/yarn/blob/master/src/package-fetcher.js).

**Comment by oliverjam** [comment]
@ronanyeah Those workshops look good. I was envisioning a promises/fetch workshop beginning with XMLHttpRequest and then showing how you could wrap it to return a promise (creating your own version of fetch). It would be a nice way to intro all the different bits and tie them together, and could end with "nice browsers have done this for you with the fetch API".

I was going to mention that this opens the door to async/await, but thought that might have been a step too far 😉

**Comment by ronanyeah** [comment]
@oliverjam Yeah that's basically what they are. The XHR one wraps XMLHttpRequest with callbacks, and the Promise one is for wrapping it with Promises. It's just QUnit tests that need to be passed.

**Comment by jsms90** [comment]
@oliverjam Thank you for raising this. I completely agree that the WebRTC workshop is deeply problematic for anyone who hasn't had an introduction to the kinds of promises (also classes and OOP conceptually). I also think promises should be in this course!

But, you might notice a pattern soon, I think we should tackle the problem of introducing callbacks _well_ first. @ronanyeah how do you do it in gaza?

**Comment by ronanyeah** [comment]
I wrote two workshops. :)

**Comment by jsms90** [comment]
Currently being introduced into the course via [a research topic](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/week-8/research-afternoon.md#promises-and-fetch) in week 8

**Comment by zurda** [comment]
@jsms90 , we have a PR suggesting to [add callbacks as a research topics in week 2](https://github.com/foundersandcoders/master-reference/pull/961), which might make things a bit less confusing when they get to promises.
For now, as far as I can see - promises were added as a research topic and a morning challenge, so I'm not sure if the discussion is still active, or if this is considered resolved?

**Comment by oliverjam** [comment]
I'm honestly not sure. This issue is so old, but a lot of what I originally raised is still an issue. If anything it's worse now—there are _tons_ of browser APIs that use promises (fetch, cache, geolocation, bluetooth, permissions etc). As of [two days ago](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md#11.14.0) Node core has stable support for promise-based file system access (`const fs = require("fs").promises`)

We're in danger of being left behind a bit here—I think at this point in _every_ group in FAC16 has ended up using promises in at least one project, and they still haven't been officially mentioned or "allowed" by the curriculum.

It's actually difficult to try and keep async callbacks around—see [my issue](https://github.com/foundersandcoders/master-reference/issues/978) about the `request` module workshop. The module has been deprecated in favour of other promise-based ones, so what do we recommend now?

I don't know if a single morning challenge in Week 7 is enough given how much time we end up devoting to callbacks. I also don't see callbacks and promises as competing for mindshare with students. You still need to understand callbacks to understand promises, so callbacks continue to get reinforced regardless.

**Comment by zurda** [comment]
I might be missing something here, and I apologise in advance, but it would be helpful if we focus the topic a bit. My confusion is - is there an agreement that promises should be introduced earlier? 

I can see two problems raised in this issue: 
1. Promises is introduced in a morning challenge in week-7, yet the research topic about promises is only a week later + promises is a complex topic and should be better introduced.  
2. We should introduce promises earlier, ideally in API week.  

I would be a huge fan of teaching promises earlier, but I can't figure out from the issue if everyone's on board and I think it's important to clarify that before moving forward. 

If there's an agreement that (2) is an issue and we should teach promises earlier, we might not even bother fixing (1). But if (2) is still in discussion, then fixing problem 1 might help in the meantime. 

I think my confusion might be a cause of being told I'm not allowed to use promises prior to week 7. It felt like there was some reasoning to it, some lesson I need to learn. I apologise again if I misread something (language barrier excuse! :) ), but I would appreciate any clarification about this topic.

**Comment by oliverjam** [comment]
So I originally opened this issue during curriculum planning for FAC11. We (FAC10) didn't cover promises at all, so it was confusing for us to emerge from the first 8 weeks or so and encounter them for the first time. Since then (possibly some time around FAC13) a morning challenge was created to introduce promises a bit, during Week 7. I'm not sure when the research afternoon was added, but that's as far as we've gotten.

Personally I don't think a morning challenge in Week 7 is enough for such an important part of the JS language, but I have no idea where the consensus is on potentially having more promise-based material earlier in the course 🤷‍♂️

**Comment by arrested-developer** [comment]
May be related to #971 - I think @m4v15 introduced fetch and thus promise syntax in week 2 for FAC-K1 , although promises weren't fully explored until later in the course.

**Comment by m4v15** [comment]
Yeah so like @arrested-developer says we kind of introduced promises in week 2, but just the syntax really as we still wanted people to understand the callback pattern - something which I still think _has_ to be understood well before moving straight onto callbacks

I think it's definitely fair to be confused by this @zurda, how we introduce asynchronous javascript is a huge issue that has never been properly addressed, and has been raised a lot, probably most comprehensively discussed in #874 

I think the main thing is we probably all agree that promises could be introduced earlier but they are obviously connected to how we introduce async as a whole, and I think that is a the larger issue that needs to be addressed?
